### ISSUE TITLE
Should we have built-in support for conditional productions, stochastic productions?

### ISSUE THREAD

**Issue by TMiguelT** [nylki/lindenmayer#10]
Hi! You've done fantastic work on this library so far. However I've been reading through ABOP and I wonder if the API could support some of the standard notational things in the book.

For example, stochastic productions and conditional productions seem very common, so could we have additional keys on the production object for these? For example:

```javascript
lsys.setProduction('F', 'F-F++F-F', {chance: 0.3, condition: () => time > 3})
```

**Comment by nylki** [comment]
Hi @TMiguelT,
Thanks!
First of all, your example above could already been written as:
```.js
lsys.setProduction('F', () => { if (Math.random() <= 0.3 && time > 3) return  'F-F++F-F'})
```

But I see your point about providing better readability for the common use case you mentioned (probability and some condition).
I have been thinking about this for a while before.
Let me explain my hesitation to implement it like you proposed:
In ABOP all productions are set together in the beginning with a defined order.
So you have eg. 3 productions: first one with 25% probability, second also 25% and the third with 50%.
Now the issues I see that arise:

⁻ If we would adapt it for our library, we'd have to take into consideration that percentages don't match up. What do we do if percentages are not exactly 100% but more, because of users miscalcuations. Do we Just cap/fill the ast productions probability? This is probably less serious as we could have a console warning.

- But the more difficult thing in my opinion: What happens when we want to add more productions later on? We'd have to either set all previous productions again and calculate our probability percentages to match eg. 4 or 5 productions. Otherwise, as our 3 example productions already reach 100% total probability the additionally added productions would never have a chance to be executed because of order of execution from top to bottom. 

Conclusion: This approach would work when you don't modify productions after setting them in the beginning. Or when all probabilities are equal, so adding more production would automatically change the probabilites of the others. However that would defeat the purpose of the `chance` option.

Do you see my point here? :)
The chance notation makes sense in a formal definition as in ABOP, but I don't think it is that practical in the library.
However, as I would really like to have more readability and adaptability to ABOP examples I am open to suggestions!

**Comment by nylki** [comment]
Ok, I skimmed the stochastic LSystem section of ABOP quickly again. The *first* implementation appears to be the correct one as I remembered.
What migh be possible is simply to introduce the option as part of the classic/ABOP syntax package which I have already added to some degree (context sensitivity with `< and >`) and simply make it clear that when using `chance`, you should be careful when using `setProduction` or `setProductions` on ABOP-stochastic productions. And emit a warning/error when probability is over 100%.

**Comment by TMiguelT** [comment]
I think you're correct in the first interpretation. The way I'd implement this is with production 'groups' that have a number of productions with weights. It sounds like you've started doing something like this over in #9, but I think adding weights to each production is a nice simple way of setting a chance for each production without forcing the user to add them to 100%. For example:
```javascript
let lsystem = new LSystem({
      axiom: 'ABC',
      productions: {
        'A': 'A+',
        'B': [
             {weight: 3, succ: 'AB'}, // 60%
             {weight: 2, succ: 'BA'} // 40%
        ],
        'C': 'ABC'
      }
})
```

**Comment by TMiguelT** [comment]
I realise you're trying to make the ABOP notation easy to write literally in JS, but I admit I prefer a more JavaScripty API where each production is an object with a succ (successor string or function), condition (function), left_context (string), right_context (string) etc. Rather than using strings for everything. I especially like how you've implemented parametric systems, because they're very JavaScripty and intuitive, even if you haven't read papers on L-Systems.

**Comment by nylki** [comment]
I like your idea using weights.  Will think about how that could be implemented including other often used operations like context checks in a objecty way as you suggest. I still want it to be pretty flexible, so for people who want to do basic L-Systems or want to do using only strings or want to define a function or other object for every production it should work as before.

Talking with myself: What about in `setProduction`, check if production is an object or list (done already), then also check for objects whether keywords like `successor` or `leftContext` are included and for lists if multiple objects with weights are inside; Then: construct functions that perform as expected (as it is already done when transforming classic context sensitive syntax to functions).

**Comment by TMiguelT** [comment]
With my ideal API you'd allow almost everything to be an object (how I'd probably use it) or a string (the classic ABOP way), or some other things.

With this API, everywhere you have a predecessor (left hand side), it can be:
* A string, in which case you interpret in the classical way, meaning it can have a symbol and optionally a left and right context.
* An object, which would have `{symbol, left_ctx, right_ctx}` values

If it's a string, you can parse it into an object, so you can always treat the predecessor as an object.

Everywhere you have a successor, it can be:
* A string, in which case it's a simple conversion e.g. `F->FF`
* A function, in which case they can use parameters on the predecessor
* An object, which would have the keys `{chance (number), condition (function), successor (function or string)}`
* An array of objects as above, except they can have a `weight` key as well.

So you'd check which type the successor is, and then if it's a string, you make that into an object with just the successor key, and then put that in an array by itself (return `[{successor: x, weight: 1}]`. If it's a function, you do the same (`return [{successor: x, weight: 1}]`). If it's an object you put that in an array with a weight of 1 (`return [Object.assign({weight: 1}, x)]`), and then you can *always* treat the successors as an array of objects, which saves you having to switch case everywhere, except on the successor key which can always be a string or a function.

This way, your code should be pretty straight forward since no matter what the user does you can write your code for one data structure. Also, if you provide user documentation showing the options they have like I've done above, it should be pretty clear how to use your module.

**Comment by nylki** [comment]
I generally like your suggestions so far but I see a few problems we'd have to tackle or take into consideration.

**1. simplicity**

> If it's a function, you do the same (return [{successor: x, weight: 1}])

This is fine if the user wants to use/return objects anyways. But I really love the simplicity of something like the following when using ES6 arrow functions:
```.js
// This:
setProduction('F', () => (time > 2) ? 'FF' : 'F-')

// would have to become this, even though the user doesnt want to use objects and only use functions for dynamic conditions. according to your suggestion this would become:
setProduction('F', () => {
    return { successor: (time > 2) ? 'FF' : 'F-', weight: 1 }
})
```

I don't want to sacrifice the ability to do it the first way. Therefore I propose that Strings should remain to always be allowed as a functions result. To be able to, we'd check wether a result is an object or a string before appending it to the successor axiom/word. If it is indeed a string we could transform it to a normalized object in the form of  `is result a string then return {sucessor: result, weight: 1} otherwise return result`. This would certainly sacrifice processing power, but we could do some benchmarks to see the actual impact.

**2. Limits of JSON / defining productions in the constructor**
Currently you can set productions like: 
```.js
let lsystem = new LSystem({
	productions: {
		'F': [{symbol: 'F'}, {symbol: 'F'}],
		'C': {symbol: 'G'}
	}
})
```

We **use an JSON-object** to define productions, which gets parsed into a Map for easy and fast production retrieval. However JSON-objects don't allow objects to be used as keys, only strings are allowed on the left side.
What you propose, having objects on the left side, would not work for the above way, because:

```.js
let lsystem = new LSystem({
	productions: {
                /* error, invalid JSON object: */
		{predecessor: 'F'}: {successor: [{symbol: 'F'}, {symbol: 'F'}]},
		{predecessor: 'C'}: {sucessor: {symbol: 'G'}}
	}
})
```
would already throw errors because you'd try to use objects for keys in the object (left side).
Which means that we'd have to use either the existing function (setProduction) or change it to be an array with tuples/two-value-arrays, instead of an object with key:values.

```.js
let lsystem = new LSystem({
	productions: [
		[ {predecessor: 'F'}, {successor: [{symbol: 'F'}, {symbol: 'F'}]} ],
		[ {predecessor: 'C'}, {successor: {symbol: 'G'}} ]
	]
})
```
The above would work, but is certainly a bit less concise than it used to be.

**3. memory**
Treating each symbol as an individual object uses certainly more memory than using pure strings. Surely this would already be a problem when not using basic strings and doing parametric stuff. But it would add more overhead for simple situations when you want to only use strings to begin with. A benchmark to see the actual memory and performance impact in string-only-situations may be useful here.

**Conclusion**
After thinking about the above points I suggest the following modifications to your proposal:
I see that you want to keep `leftCtx` and `rightCtx` on the left side similiar to ABOPs definition (`A<B>C -> A+A`).
What do you think about keeping a single String on the left side as it is now and defining all contexts, conditions, weights on the right side. This would the parsing a bit more straight forward and would keep the conciser construction (see 2.).

So a new workflow to set production may look like:

```.js
setProduction('F', {
	/* set a basic condition*/
	condition: () => time > 1.0,
	
	stochasticSuccessors: [
		{
			weight: 0.2,
			successor: [{symbol: 'F', customParameter: 2}, {symbol: '+'}, {symbol: 'F', customParameter: 5}]},
		{
			weight: 0.3,
			successor: {symbol: 'F', customParameter: 0}},
		{
			weight: 0.5,
			/* additional condition in a case when this sucessor is choosen with a chance of 50%*/
			condition: () => time > 9000.0,
			successor: [{symbol: 'G', customParameter: 2}, {symbol: 'F', customParameter: 2}, {symbol: '-'}]}
	]
});


setProduction('B', {
	/* set a basic condition*/
	condition: () => time > 1.0,
	leftCtx: 'FFB',
	rightCtx: '+B',
	sucessor: 'BB' /* which would internally get transformed to the equivalent object: {symbol: 'BB', weight: 1}*/
});

// Or with explicit classicSyntax option this may be allowed:

setClassicProduction('FFB<B>+B', {
	/* set a basic condition*/
	condition: () => time > 1.0,
	sucessor: 'BB' /* which would internally get transformed to the equivalent object: {symbol: 'BB', weight: 1}*/
});

//or something like: 

setProduction('FFB<B>+B', {
	allowClassicSyntax: true,
	/* set a basic condition*/
	condition: () => time > 1.0,
	sucessor: 'BB' /* which would internally get transformed to the equivalent object: {symbol: 'BB', weight: 1}*/
});

// Or when defining it together via the constructor:

let lsystem = new LSystem({
	axiom: [{symbol: 'F'}, {symbol: '+'}, {symbol: 'F'}],
	productions: {
		
		'F': {
			/* set a basic condition*/
			condition: () => time > 1.0,
			leftCtx: 'FFB',
			rightCtx: '+B',
			sucessor: 'BB' /* which would internally get transformed to the equivalent object: {symbol: 'BB', weight: 1}*/
		},
		
		'B': {
			leftCtx: 'F+',
			stochasticSuccessors: [
				{ weight: 0.2, successor: [{symbol: 'F', customParameter: 2}, {symbol: '+'}, {symbol: 'F', customParameter: 5}]},
				{ weight: 0.3, successor: {symbol: 'F', customParameter: 0}},
				{
					weight: 0.5,
					condition: () => time > 9000.0,
					successor: [{symbol: 'G', customParameter: 2}, {symbol: 'F', customParameter: 2}, {symbol: '-'}]}
			]
		}
		
	}
});


```

**Comment by nylki** [comment]
I kinda also like your initial idea `setProduction(from, condition, to)` which is really nice in this form because it nicely separates everything in a logical way. But this would mess again with setting the prods in the constructor and would require using an array of triples.

**Comment by TMiguelT** [comment]
Oh sorry I'm not sure I made myself clear. What I meant was that the user can specify the production as a:
* string
* object
* array
* function

But lindenmayer.js, *internally* would convert all of these to arrays of objects. But looking at what you've suggested, I much prefer using objects for all the successors, with an optional `stochasticSuccessors` array. That looks really nice. And I entirely agree with your proposal to put the left and right context on the RHS object, your reasoning is good (makes it useable with dictionaries etc.)

***

So, having taken that into account, here's my revised proposal:

Everywhere you can set a production (in the LSystem constructor, in `setProduction` and `setProductions`), there will be a LHS (the dictionary key or first argument), and the RHS (the dictionary value or second argument)

The LHS must be a string containing the predecessor symbol, optionally with a left and right context in the classical syntax. These left and right contexts will be converted to `leftContext` and `rightContext` on the successor object.

The RHS can be:
* A string, which is *internally* converted to `{successor: x}`, where x is the string. Note, if this causes memory issues, we *could* keep this as a string, and then switch case every time we need to run a production. This would improve the performance but messy the internal code somewhat
* A function, which is internally converted to `{successor: x}`, where x is the function  
* An object, which can have the keys:
    * `leftCtx`: the left context
    * `rightCtx`: the right context
    * `condition`: a function with the signature object -> boolean that returns true if we should use this production, otherwise false
    * `successor`, a string or function returning a string that determines the successor symbols
    *  `weight`, only relevant if part of a `stochasticSuccessors` array. Determines the chance that this production is used
    * `stochasticSucessors`: an array of more RHS objects in any of the three types described here (string, function, object), from which one production will be chosen, based on `weight`. 

Thoughts?

**Comment by TMiguelT** [comment]
Actually why can't we use the same key for stochastic successors and normal successors? Then we just say that successor can be either:
* A string
* A function
* An array of objects

Then, they can set the RHS as any of these options, which would all be internally converted to an object, or they can pass a full object containing any of the keys above

**Comment by nylki** [comment]
> Actually why can't we use the same key for stochastic successors and normal successors? Then we just say that successor can be either

It would work of course. It's just a matter of style.
I kind of like to have objects x in `successor: x` to be of the same format in all places. To me it feels irritating to have key:values likes `weight` and other `successor`s inside an object that is called `successor` which indicates that this would be the final form in which it is going to be stored in memory. Having the separate identifier makes the distinction better pronounced in my opinion.

Also you could have an array of functions btw. (that return either `false` in which case the next function or value in the array is used or they return a valid successor). See this line: https://github.com/nylki/lindenmayer/blob/master/test/tests.js#L155

**Comment by nylki** [comment]
> Thoughts?

I like it! 👍 
But what do you think of the alternative, similiar to your initial idea:

```.js

// Scaffold all conditions, stochastic stuff and options into a third argument
setProduction('F', {leftCtx: '+G', condition: someFc }, 'FFF');

// works also with only two arguments
setProduction('F', 'FFF');

setProduction('F', {leftCtx: '+G', condition: someFc }, [{symbol: 'F'}, {symbol: 'G'}]);

setProduction('F', {leftCtx: '+G', condition: someFc }, {
	stochasticSuccessor: [
		{weight: 0.2, successor: [{symbol: 'F'}, {symbol: 'G'}, {symbol: 'G'}] },
		{weight: 0.8, successor: [{symbol: 'G'}, {symbol: 'G'}] }
	]
});


// And when defining in constructor the same as we discussed.
let lsys = new LSystem({
	axiom: 'F',
	productions: {	
		'F': { leftCtx: '+G',  condition: someFc, successor: {symbol: 'G'} }
		'G': { leftCtx: '-FF', condition: someFc, successor: {symbol: 'G'} },
		'H': 'FF'
	}
})
```

**Comment by TMiguelT** [comment]
I think I prefer only two sections to the arguments (LHS and RHS). Ideally I'd only want 1 section, but that stops us using the nice dictionary syntax so 2 is a good compromise

**Comment by nylki** [comment]
I'm working on something based on our ideas right now. I'll let you know (in ~1-3 weeks) when you can test and play around with it!

**Comment by nylki** [comment]
Quick heads up, I am still on it. Got sick and couldn't do anything productive the past two weeks :/

**Comment by nylki** [comment]
closed via #16
